File: problems/ruby/0002_Add_Two_Numbers.rb
# Definition for singly-linked list.
class ListNode
  attr_accessor :val, :next

  def initialize(val)
    @val = val
    @next = nil
  end
end

# @param {ListNode} l1
# @param {ListNode} l2
# @return {ListNode}
def add_two_numbers(l1, l2)

  node = ListNode.new(0)
  ret = node
  carry = 0
  until l1.nil? && l2.nil?
    val1 = l1.nil? ? 0 : l1.val
    val2 = l2.nil? ? 0 : l2.val
    x = val1 + val2 + carry
    if x >= 10
      carry = 1
      x %= 10
    else
      carry = 0
    end
    node.next = ListNode.new(x)
    node = node.next
    l1 = l1.next unless l1.nil?
    l2 = l2.next unless l2.nil?
  end
  node.next = ListNode.new(1) if carry == 1
  ret.next
end

File: problems/go/0002_Add_Two_Numbers.go
package _go

type ListNode struct {
	Val  int
	Next *ListNode
}

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {

	var l3 *ListNode
	ret := l3
	carry := 0

	for l1.Next != nil || l2.Next != nil {
		sum := l1.Val + l2.Val
		if sum > 9 {
			carry = 1
			sum %= 10
		} else{
			carry = 0
		}
		l3.Val = sum + carry
		l3.Next = *ListNode
		l3 = l3.Next
		l1 = l1.Next
		l2 = l2.Next
	}

	return ret
}
